clippy_lints/src/ptr.rs

```diff
@@ -111,6 +128,28 @@ fn check_fn(cx: &LateContext, decl: &FnDecl, fn_id: NodeId) {
             }
         }
     }
+
+    if let FunctionRetTy::Return(ref ty) = decl.output {
+        if let Some((out, MutMutable)) = get_rptr_lm(ty) {
+            if let Some(MutImmutable) = decl.inputs.iter()
+                     .filter_map(|ty| get_rptr_lm(ty))
+                     .filter(|&(lt, _)| lt.name == out.name)
+                     .fold(None, |x, (_, m)| match (x, m) {
+                        (Some(MutMutable), _) |
+                        (_, MutMutable) => Some(MutMutable),
+                        (_, m) => Some(m),
+                     }) {
+                span_lint(cx,
+                          MUT_FROM_REF,
+                          ty.span,
+                          "this function takes an immutable ref to return a mutable one:");
```

Comment:
OK, how do I get a span of the function header (e.g. from the `fn` to just before the opening bracket or semicolon, minus whitespace)?